File: internal/squirrel/sink_test.go
package squirrel_test

import (
	"context"
	"errors"
	"testing"

	"github.com/stretchr/testify/require"

	"github.com/ronaldlokers/squirrel/internal/squirrel"
)

type brokenWriter struct{}

func (brokenWriter) Write(squirrel.Capture) (string, error) {
	return "", errors.New("no space left on device")
}

type panickingWriter struct{}

func (panickingWriter) Write(squirrel.Capture) (string, error) {
	panic("writer exploded")
}

func TestSinkStoresAnAcceptedCapture(t *testing.T) {
	sp, _ := openSpool(t)
	sink := squirrel.NewSink(sp, allows, squirrel.SinkHooks{})

	require.Equal(t, squirrel.Stored, sink.Accept(context.Background(), capture(nil)))

	names, err := sp.List()
	require.NoError(t, err)
	require.Len(t, names, 1)
}

func TestSinkIgnoresAndWritesNothing(t *testing.T) {
	sp, _ := openSpool(t)
	sink := squirrel.NewSink(sp, allows, squirrel.SinkHooks{})

	c := capture(func(c *squirrel.Capture) { c.ConversationID = squirrel.Ptr("8") })
	require.Equal(t, squirrel.Ignored, sink.Accept(context.Background(), c))

	names, err := sp.List()
	require.NoError(t, err)
	require.Empty(t, names)
}

// Silence in the room, but not silence in the logs.
func TestSinkReportsAnIgnoredConversation(t *testing.T) {
	sp, _ := openSpool(t)
	var seen *squirrel.Capture
	sink := squirrel.NewSink(sp, allows, squirrel.SinkHooks{
		OnIgnored: func(c squirrel.Capture) { seen = &c },
	})

	c := capture(func(c *squirrel.Capture) { c.ConversationID = squirrel.Ptr("8") })
	sink.Accept(context.Background(), c)

	require.NotNil(t, seen)
	require.Equal(t, "8", *seen.ConversationID)
}

func TestSinkStoresAFailOpenCapture(t *testing.T) {
	sp, _ := openSpool(t)
	sink := squirrel.NewSink(sp, allows, squirrel.SinkHooks{})

	c := capture(func(c *squirrel.Capture) {
		c.ConversationID = nil
		c.SenderID = nil
	})
	require.Equal(t, squirrel.Stored, sink.Accept(context.Background(), c))
}

// The one path where the bot admits it could not do its job.
func TestSinkReportsFailureRatherThanPretending(t *testing.T) {
	var got error
	sink := squirrel.NewSink(brokenWriter{}, allows, squirrel.SinkHooks{
		OnError: func(err error) { got = err },
	})

	require.Equal(t, squirrel.Failed, sink.Accept(context.Background(), capture(nil)))
	require.Error(t, got)
}

// A panicking writer must not take the process down with it. Accept always
// returns an Outcome.
func TestSinkSurvivesAPanickingWriter(t *testing.T) {
	sink := squirrel.NewSink(panickingWriter{}, allows, squirrel.SinkHooks{})
	require.Equal(t, squirrel.Failed, sink.Accept(context.Background(), capture(nil)))
}

// A hook is observability. It must not be able to break the path it observes.
func TestSinkSurvivesAPanickingHook(t *testing.T) {
	sp, _ := openSpool(t)
	sink := squirrel.NewSink(sp, allows, squirrel.SinkHooks{
		OnIgnored: func(squirrel.Capture) { panic("hook exploded") },
	})

	c := capture(func(c *squirrel.Capture) { c.ConversationID = squirrel.Ptr("8") })
	require.Equal(t, squirrel.Ignored, sink.Accept(context.Background(), c))

	names, err := sp.List()
	require.NoError(t, err)
	require.Empty(t, names)
}

File: test/unit/sink.test.ts
import { mkdtemp } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { beforeEach, describe, expect, it, vi } from "vitest";

import type { Capture } from "../../src/capture/capture.js";
import type { Allow } from "../../src/capture/policy.js";
import { createSink } from "../../src/capture/sink.js";
import { type Spool, openSpool } from "../../src/capture/spool.js";

const allows: readonly Allow[] = [{ transport: "campfire", conversationId: "7", senderId: "1" }];

let spool: Spool;

beforeEach(async () => {
  spool = await openSpool(await mkdtemp(join(tmpdir(), "squirrel-sink-")));
});

function capture(overrides: Partial<Capture> = {}): Capture {
  return {
    transport: "campfire",
    externalId: "42",
    conversationId: "7",
    senderId: "1",
    text: "buy milk",
    receivedAt: new Date("2026-08-14T09:31:04.512Z"),
    payload: {},
    ...overrides,
  };
}

describe("createSink", () => {
  it("stores an accepted capture", async () => {
    const sink = createSink(spool, allows);
    expect(await sink.accept(capture())).toBe("stored");
    expect(await spool.list()).toHaveLength(1);
  });

  it("ignores a capture from elsewhere and writes nothing", async () => {
    const sink = createSink(spool, allows);
    expect(await sink.accept(capture({ conversationId: "8" }))).toBe("ignored");
    expect(await spool.list()).toEqual([]);
  });

  // Silence in the room, but not silence in the logs. Being added to a room
  // Squirrel does not belong in should be visible somewhere.
  it("reports an ignored conversation so it is not silent everywhere", async () => {
    const onIgnored = vi.fn();
    const sink = createSink(spool, allows, { onIgnored });

    await sink.accept(capture({ conversationId: "8" }));

    expect(onIgnored).toHaveBeenCalledOnce();
    expect(onIgnored.mock.calls[0]?.[0]).toMatchObject({ conversationId: "8" });
  });

  // A hook is observability. It must never be able to turn a real outcome
  // into a thrown rejection that a transport then misreports to the user.
  it("still ignores and writes nothing when onIgnored throws", async () => {
    const onIgnored = vi.fn(() => {
      throw new Error("logging backend unreachable");
    });
    const sink = createSink(spool, allows, { onIgnored });

    expect(await sink.accept(capture({ conversationId: "8" }))).toBe("ignored");
    expect(await spool.list()).toEqual([]);
  });

  it("stores a fail-open capture", async () => {
    const sink = createSink(spool, allows);
    expect(await sink.accept(capture({ conversationId: null, senderId: null }))).toBe("stored");
    expect(await spool.list()).toHaveLength(1);
  });

  // The one path where the bot admits it could not do its job. Inventing a
  // cheerful answer here is exactly the false trust the design rules out.
  it("reports failure rather than pretending, when the spool cannot write", async () => {
    const broken: Spool = {
      ...spool,
      write: () => Promise.reject(new Error("ENOSPC: no space left on device")),
    };
    const onError = vi.fn();
    const sink = createSink(broken, allows, { onError });

    expect(await sink.accept(capture())).toBe("failed");
    expect(onError).toHaveBeenCalledOnce();
  });

  // Same guarantee on the failure path: a throwing onError must not mask the
  // fact that the capture was not saved.
  it("still reports failure when onError itself throws", async () => {
    const broken: Spool = {
      ...spool,
      write: () => Promise.reject(new Error("ENOSPC: no space left on device")),
    };
    const onError = vi.fn(() => {
      throw new Error("logging backend unreachable");
    });
    const sink = createSink(broken, allows, { onError });

    expect(await sink.accept(capture())).toBe("failed");
  });
});
